consensus/istanbul/backend/backend_test.go

```diff
@@ -311,42 +706,84 @@ func Test_GossipSubPeerTargets(t *testing.T) {
+			valSet.CalcProposer(valSet.GetProposer().Address(), uint64(round))
 
-		// Test if the testing node is in the committee in specified blocks and target node list is returned
-		if (checkInCommitteeBlocks(i) && targets == nil) || (!checkInCommitteeBlocks(i) && targets != nil) {
-			t.Errorf("Wrong committee information. Block: %d, Nil targets: %v\n", i, targets == nil)
-		}
+			// Use block number as prevHash. In SubList() only left 15 bytes are being used.
+			hex := fmt.Sprintf("%015d000000000000000000000000000000000000000000000000000", i)
+			prevHash := common.HexToHash(hex)
+
+			// current and next committee
+			committees := make([][]istanbul.Validator, 2)
+
+			viewCurrent := backend.currentView.Load().(*istanbul.View)
+			committees[0] = valSet.SubList(prevHash, viewCurrent)
+
+			viewCurrent.Round = viewCurrent.Round.Add(viewCurrent.Round, common.Big1)
+			backend.currentView.Store(viewCurrent)
+
+			valSet.CalcProposer(valSet.GetProposer().Address(), uint64(round+1))
+			committees[1] = valSet.SubList(prevHash, viewCurrent)
+
+			// Reduce round by 1 to set round to the current round before calling GossipSubPeer
+			viewCurrent.Round = viewCurrent.Round.Sub(viewCurrent.Round, common.Big1)
+			valSet.CalcProposer(valSet.GetProposer().Address(), uint64(round))
+			backend.currentView.Store(viewCurrent)
 
-		if checkInCommitteeBlocks(i) {
-			// targets don't have testing node's address because a target is a receiving node
-			if len(targets) == len(council)-1 {
-				for _, x := range council {
-					// council[0] is testing node and it should be removed from targets
-					if x == council[0] {
-						continue
+			targets := backend.GossipSubPeer(prevHash, valSet, nil)
+
+			// Check if the testing node is in a committee
+			check1 := backend.checkInSubList(prevHash, valSet)
+			check2 := checkInCommitteeBlocks(i, round)
+
+			if check1 != check2 {
+				t.Errorf("Difference in expected data and calculated one. Changed committee selection? HARD FORK may happen!! Sequence: %d, Round: %d", i, round)
+			} else {
+				if check1 == false {
+					continue
+				}
+			}
+
+			if len(targets) <= len(committees[0])+len(committees[1]) {
+				for n := 0; n < len(committees); n++ {
+					for _, x := range committees[n] {
+						if _, ok := targets[x.Address()]; !ok && x.Address() != backend.Address() {
+							t.Errorf("Block: %d, Round: %d, Committee member %v not found in targets", i, round, x.Address().String())
+						} else {
+							targets[x.Address()] = false
+						}
 					}
+				}
 
-					if _, ok := targets[x]; !ok {
-						t.Errorf("Gossipping target doesn't match")
+				for k, v := range targets {
+					if v == true {
+						t.Errorf("Block: %d, Round: %d, Validator not in committees included %v", i, round, k.String())
 					}
 				}
 			} else {
-				t.Errorf("Gossipping target doesn't match")
+				t.Errorf("Target has too many validators. targets: %d, sum of committees: %d", len(targets), len(committees[0])+len(committees[1]))
 			}
```

Comment:
It is hard for me to follow the logic of the test. It would be better if this can be simplified.

Reply:
@ehnuje I added more comments and split unnecessarily nested if statement. Please take a look. Thanks.